filingcabinet/views.py

```diff
@@ -178,6 +180,50 @@ def get_context_data(self, **kwargs):
         return context
 
 
+class DocumentCollectionZipDownloadView(AuthMixin, PkSlugMixin, DetailView):
+    model = DocumentCollection
+    redirect_url_name = "filingcabinet:document-collection_zip"
+    redirect_short_url_name = "filingcabinet:document-collection_zip_short"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        context.update(get_document_collection_context(self.object, self.request))
+        return context
+
+    def render_to_response(self, context):
+        archive_stream = zipstream.ZipFile(mode="w")
+        coll_docs = CollectionDocument.objects.filter(collection=self.object)
+        directory_dirname_map = {}
+        for doc in coll_docs:
+            if doc.directory is not None:
+                if doc.directory.pk not in directory_dirname_map:
+                    path_to_root = [
+                        *(x.name for x in doc.directory.get_ancestors()),
+                        doc.directory.name,
+                    ]
+                    dirname = "/".join(path_to_root)
+                    directory_dirname_map[doc.directory.pk] = dirname
+                else:
+                    dirname = directory_dirname_map[doc.directory.pk]
+                filename = "/".join(
+                    [
+                        dirname,
+                        doc.document.get_document_filename(),
+                    ]
+                )
+            else:
+                filename = doc.document.get_document_filename()
+            archive_stream.write(doc.document.get_file_path(), arcname=filename)
+
+        resp = StreamingHttpResponse(
```

Comment:
This will accumulate the whole zip tree structure in memory and only start streaming once the `coll_docs` is iterated over. I guess this is ok, but I'm wondering if the iteration over the `coll_docs` could happen via [`.iterator()`](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#django.db.models.query.QuerySet.iterator) and somehow bytes written to the zip file are streamed while iterating over the collection docs?

But there are likely other problems: It could be that nginx will kill this after 30 seconds even if bytes are flowing (might have to check on this).

[This could be part of an alternative solution](https://github.com/travcunn/django-zip-stream), ideally nginx would build and serve the ZIP file while still receiving info on which files it will contain but I'm also not sure about that. But I like the idea of offloading work from the app server. (I don't like having to build our own nginx for that, but might be worth it.)

Reply:
I also stumbled upon mod_zip, but decided that adding more dependency on using nginx might not be what we want, especially since this is not a module included in nginx by default.

I *think* nginx should let the connection live as long as bytes are flowing (proxy_read_timeout only applies for the time between successful reads, not the entire request).

I also tried to generate the zip file structure on the fly using https://github.com/uktrade/stream-zip, but always ended up with `SynchronousOnlyOperation`-Exceptions, because then the coll_docs iteration ends up within the async asgi handler context (https://code.djangoproject.com/ticket/32798)

Reply:
The fix for this is now [merged in the django project](https://github.com/django/django/pull/16384), but will only be added in django 4.2. Until then, I'd suggest to keep this solution